pkg/sidecar/pod.go

```diff
@@ -33,8 +35,39 @@ const (
 
 // add a new sidecar container to the given pod, based on the given OpenTelemetryCollector.
 func add(cfg config.Config, logger logr.Logger, otelcol v1alpha1.OpenTelemetryCollector, pod corev1.Pod, attributes []corev1.EnvVar) (corev1.Pod, error) {
-	// add the container
 	volumes := collector.Volumes(cfg, otelcol)
+	volumes[0] = corev1.Volume{
+		Name: naming.ConfigMapVolume(),
+		VolumeSource: corev1.VolumeSource{
+			EmptyDir: &corev1.EmptyDirVolumeSource{},
+		},
+	}
+	volumeMount := corev1.VolumeMount{
+		Name:      naming.ConfigMapVolume(),
+		MountPath: "/conf",
+	}
+
+	otelColCfg, err := reconcile.ReplaceConfig(otelcol)
+	if err != nil {
+		return pod, err
+	}
+	otelColCfgStr := base64.StdEncoding.EncodeToString([]byte(otelColCfg))
+
+	// add the container
+	pod.Spec.InitContainers = append(pod.Spec.InitContainers, corev1.Container{
```

Comment:
@jaronoff97 @Aneurysm9 could you please take a look and give me 👍🏼 on the approach?

The approach was chosen to avoid copying otelcol config map to the namespace where the injection is enabled and as well avoid issues with cleaning the config map. 

My notes:
* owner reference cannot reference objects from a different namespace
* when a pod is created directly it does not have any owner references

Once we agree I will make the init container image configurable via a flag on the operator (if needed expose it in the CR as well).

Reply:
Can you expand on the concerns with using config maps?  While the proposed system appears to work, it's got more moving parts than I like.

Reply:
> Can you expand on the concerns with using config maps? 

Correctly implementing the config map approach for the sidecar is tricky:
* The sidecar config map would have to be created in the target namespace.
* A custom cleaning would have to be implemented.  An owner reference to `otelcol` cannot be used (it's a cross namespace) nor owner reference to a pod (a pod UID is not known in the admission webhook).


The init container approach is a cleaner approach because no cleaning is required and the exact sidecar configuration is known (via the `OTEL_CONFIG` env var and `cat config.yaml` in the `initContainer`).  In the current implementation in the `main` branch it is hard to figure out what is the already injected sidecar configuration if the `otelcol` CR changes.


>  but I would prefer a uniform solution for internal and external namespaces. Otherwise there is the possibility that the implementations drift apart.

The init container approach that this PR introduces is used for both internal and external namespaces. There are as well tests (unit+e2e) to ensure the solution works.  The deployment/statefulset/daemonset deployment modes keep using the mounted configmap  as before - which makes perfect sense bc there is only a single version of the "instance".

Reply:
Discussed with @pavolloffay OOB and I recommend setting the configuration in an environment variable on the collector container and using the config from there by setting the argument `--config=env:OTEL_CONFIG`.  This avoids the need for an init container simply to copy the config and leaves nothing additional to clean up.